go/staking/tests/tester.go

```diff
@@ -83,51 +138,43 @@ func StakingClientImplementationTests(t *testing.T, backend api.Backend, consens
 	} {
 		t.Run(tc.n, func(t *testing.T) { tc.fn(t, backend, consensus) })
 	}
+
+	// Update tests' state for subsequent node tests (which re-use the state).
+	state.update(backend, consensus)
 }
 
 func testInitialEnv(t *testing.T, backend api.Backend, consensus consensusAPI.Backend) {
```

Comment:
The comment from https://github.com/oasislabs/oasis-core/pull/2748#discussion_r389916325 on `testInitialEnv()` becoming tautological still applies.

I don't have a strong opinion on whether we should leave it as is or perhaps:
1. Add a `firstRun` variable to the state and only run the tests in `testInitialEnv()` the first time?
2. Remove `testInitialEnv()` altogether?

Reply:
what runs before this subtest to alter the state?

Reply:
`TestNode` first runs `Staking` tests and then `StakingClient` tests:
https://github.com/oasislabs/oasis-core/blob/e7d25bf0c9c69c22de5268633182e60a948d828f/go/oasis-node/node_test.go#L238-L240

These functions in turn run `StakingImplementationTest` and `StakingClientImplementationTests` which internally run (almost) the same tests on the same state:
https://github.com/oasislabs/oasis-core/blob/e7d25bf0c9c69c22de5268633182e60a948d828f/go/staking/tests/tester.go#L40-L69

https://github.com/oasislabs/oasis-core/blob/e7d25bf0c9c69c22de5268633182e60a948d828f/go/staking/tests/tester.go#L71-L86

Hence, the need to update the state so that individual test functions (i.e. `testTransfer`, `testSelfTransfer`, ...) can work the second (and 3rd, ...) time.

Reply:
@pro-wh, ping

Reply:
ok probably go without testInitialEnv then. I'm starting to get the feeling it was there as a cheap solution before something like this `stakingTestsState` system that you're adding.

maybe in the future we'll even have a way to run tests in their own independent state

Reply:
> ok probably go without testInitialEnv then. I'm starting to get the feeling it was there as a cheap solution before something like this `stakingTestsState` system that you're adding.

Ok, agreed.

> maybe in the future we'll even have a way to run tests in their own independent state

Yes, that would be very neat.

Reply:
I ended up extracting the staking thresholds checks from `testInitialEnv()` as `testThresholds()` since they seem useful.

Please, take another look.

Reply:
testThresholds looks good